digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java

```diff
@@ -948,41 +962,43 @@ public DatabaseSessionStore(int siteId)
-                                    catchConflict(() ->
-                                        dao.insertSession(session.getProjectId(), session.getWorkflowName(), session.getSessionTime().getEpochSecond()),
-                                        "session instant=%s in project_id=%d and workflow_name=%s", session.getSessionTime(), session.getProjectId(), session.getWorkflowName()),
-                                    "project id=%d", session.getProjectId());
-                        }
-                        catch (ResourceConflictException ex) {
-                            ts.retry(ex);
-                            return null;
-                        }
-                    }
-                    else {
-                        StoredSession storedSession = dao.getSessionByConflictedNamesInternal(session.getProjectId(), session.getWorkflowName(), session.getSessionTime().getEpochSecond());
-                        if (storedSession == null) {
-                            throw new IllegalStateException("Database state error", ts.getLastException());
-                        }
-                        sesId = storedSession.getId();
+            return DatabaseSessionStoreManager.this.<T, ResourceConflictException, ResourceNotFoundException>transaction((handle, dao) -> {
+                StoredSession storedSession;
+
+                if (dao instanceof H2Dao) {
+                    catchForeignKeyNotFound(
+                        () -> {
+                            ((H2Dao) dao).upsertAndLockSession(
+                                session.getProjectId(),
+                                session.getWorkflowName(),
+                                session.getSessionTime().getEpochSecond());
+                            return 0;
+                        },
+                        "project id=%d", session.getProjectId());
+                    storedSession = dao.getSessionByConflictedNamesInternal(
+                            session.getProjectId(),
+                            session.getWorkflowName(),
+                            session.getSessionTime().getEpochSecond());
+                    if (storedSession == null) {
+                        throw new IllegalStateException(String.format(ENGLISH,
+                                    "Database state error: locked session is null: project_id=%d, workflow_name=%s, session_time=%d",
+                                    session.getProjectId(), session.getWorkflowName(), session.getSessionTime().getEpochSecond()));
                     }
                 }
-
-                StoredSession storedSession = dao.lockSession(sesId);
-                if (storedSession == null) {
-                    throw new IllegalStateException("Database state error");
+                else {
+                    // select first so that conflicting insert doesn't increment sequence of primary key unnecessarily
+                    storedSession = dao.getSessionByConflictedNamesInternal(
```

Comment:
How about issuing this query for H2 as well since MERGE in H2 seems to increment the sequence when violating foreign key constraint?

```
sql> select * from a;
ID: 1

ID: 2
(2 rows, 0 ms)
sql> merge into b (a_id) key (a_id) values (1);
(Update count: 1, 0 ms)
sql> merge into b (a_id) key (a_id) values (3);
Error: org.h2.jdbc.JdbcSQLException: Referential integrity constraint violation: "CONSTRAINT_42_1: PUBLIC.B FOREIGN KEY(A_ID) REFERENCES PUBLIC.A(ID) (3)"; SQL statement:
merge into b (a_id) key (a_id) values (3) [23506-191]
org.h2.jdbc.JdbcSQLException: Referential integrity constraint violation: "CONSTRAINT_42_1: PUBLIC.B FOREIGN KEY(A_ID) REFERENCES PUBLIC.A(ID) (3)"; SQL statement:
merge into b (a_id) key (a_id) values (3) [23506-191]
        :
sql> merge into b (a_id) key (a_id) values (2);
(Update count: 1, 1 ms)
sql> select * from b;
ID  : 7
A_ID: 1

ID  : 9
A_ID: 2
(2 rows, 1 ms)
```